File: iOS/ai/Data/AppService.swift
import Foundation

// MARK: - API Format

enum ApiFormat: String, Codable, Sendable {
    case openaiCompatible = "OPENAI_COMPATIBLE"
    case anthropic = "ANTHROPIC"
    case google = "GOOGLE"
}

// MARK: - Endpoint Rule

struct EndpointRule: Codable, Equatable, Sendable {
    let modelPrefix: String
    let endpointType: String
}

// MARK: - App Service (Provider)

final class AppService: Identifiable, Sendable {
    let id: String
    let displayName: String
    let baseUrl: String
    let adminUrl: String
    let defaultModel: String
    let openRouterName: String?
    let apiFormat: ApiFormat
    let chatPath: String
    let modelsPath: String?
    let prefsKey: String
    let seedFieldName: String
    let supportsCitations: Bool
    let supportsSearchRecency: Bool
    let extractApiCost: Bool
    let costTicksDivisor: Double?
    let modelListFormat: String
    let modelFilter: String?
    let litellmPrefix: String?
    let hardcodedModels: [String]?
    let defaultModelSource: String?
    let endpointRules: [EndpointRule]

    // Computed
    var modelFilterRegex: NSRegularExpression? {
        guard let filter = modelFilter else { return nil }
        return try? NSRegularExpression(pattern: filter, options: .caseInsensitive)
    }

    init(
        id: String,
        displayName: String,
        baseUrl: String,
        adminUrl: String = "",
        defaultModel: String,
        openRouterName: String? = nil,
        apiFormat: ApiFormat = .openaiCompatible,
        chatPath: String = "v1/chat/completions",
        modelsPath: String? = "v1/models",
        prefsKey: String = "",
        seedFieldName: String = "seed",
        supportsCitations: Bool = false,
        supportsSearchRecency: Bool = false,
        extractApiCost: Bool = false,
        costTicksDivisor: Double? = nil,
        modelListFormat: String = "object",
        modelFilter: String? = nil,
        litellmPrefix: String? = nil,
        hardcodedModels: [String]? = nil,
        defaultModelSource: String? = nil,
        endpointRules: [EndpointRule] = []
    ) {
        self.id = id
        self.displayName = displayName
        self.baseUrl = baseUrl
        self.adminUrl = adminUrl
        self.defaultModel = defaultModel
        self.openRouterName = openRouterName
        self.apiFormat = apiFormat
        self.chatPath = chatPath
        self.modelsPath = modelsPath
        self.prefsKey = prefsKey
        self.seedFieldName = seedFieldName
        self.supportsCitations = supportsCitations
        self.supportsSearchRecency = supportsSearchRecency
        self.extractApiCost = extractApiCost
        self.costTicksDivisor = costTicksDivisor
        self.modelListFormat = modelListFormat
        self.modelFilter = modelFilter
        self.litellmPrefix = litellmPrefix
        self.hardcodedModels = hardcodedModels
        self.defaultModelSource = defaultModelSource
        self.endpointRules = endpointRules
    }

    func matchesFilter(_ modelId: String) -> Bool {
        guard let regex = modelFilterRegex else { return true }
        let range = NSRange(modelId.startIndex..., in: modelId)
        return regex.firstMatch(in: modelId, range: range) != nil
    }

    func usesResponsesApi(model: String) -> Bool {
        let lower = model.lowercased()
        return endpointRules.contains { rule in
            lower.hasPrefix(rule.modelPrefix.lowercased()) && rule.endpointType == "responses"
        }
    }
}

extension AppService: Equatable {
    static func == (lhs: AppService, rhs: AppService) -> Bool { lhs.id == rhs.id }
}

extension AppService: Hashable {
    func hash(into hasher: inout Hasher) { hasher.combine(id) }
}

// MARK: - Provider Definition (JSON-serializable)

struct ProviderDefinition: Codable, Sendable {
    let id: String
    let displayName: String
    let baseUrl: String
    var adminUrl: String?
    let defaultModel: String
    var openRouterName: String?
    var apiFormat: String?
    var chatPath: String?
    var modelsPath: String?
    var prefsKey: String?
    var seedFieldName: String?
    var supportsCitations: Bool?
    var supportsSearchRecency: Bool?
    var extractApiCost: Bool?
    var costTicksDivisor: Double?
    var modelListFormat: String?
    var modelFilter: String?
    var litellmPrefix: String?
    var hardcodedModels: [String]?
    var defaultModelSource: String?
    var endpointRules: [EndpointRule]?

    func toAppService() -> AppService {
        AppService(
            id: id,
            displayName: displayName,
            baseUrl: baseUrl,
            adminUrl: adminUrl ?? "",
            defaultModel: defaultModel,
            openRouterName: openRouterName,
            apiFormat: ApiFormat(rawValue: apiFormat ?? "OPENAI_COMPATIBLE") ?? .openaiCompatible,
            chatPath: chatPath ?? "v1/chat/completions",
            modelsPath: modelsPath,
            prefsKey: prefsKey ?? "",
            seedFieldName: seedFieldName ?? "seed",
            supportsCitations: supportsCitations ?? false,
            supportsSearchRecency: supportsSearchRecency ?? false,
            extractApiCost: extractApiCost ?? false,
            costTicksDivisor: costTicksDivisor,
            modelListFormat: modelListFormat ?? "object",
            modelFilter: modelFilter,
            litellmPrefix: litellmPrefix,
            hardcodedModels: hardcodedModels,
            defaultModelSource: defaultModelSource,
            endpointRules: endpointRules ?? []
        )
    }

    static func fromAppService(_ service: AppService) -> ProviderDefinition {
        ProviderDefinition(
            id: service.id,
            displayName: service.displayName,
            baseUrl: service.baseUrl,
            adminUrl: service.adminUrl,
            defaultModel: service.defaultModel,
            openRouterName: service.openRouterName,
            apiFormat: service.apiFormat.rawValue,
            chatPath: service.chatPath,
            modelsPath: service.modelsPath,
            prefsKey: service.prefsKey,
            seedFieldName: service.seedFieldName,
            supportsCitations: service.supportsCitations,
            supportsSearchRecency: service.supportsSearchRecency,
            extractApiCost: service.extractApiCost,
            costTicksDivisor: service.costTicksDivisor,
            modelListFormat: service.modelListFormat,
            modelFilter: service.modelFilter,
            litellmPrefix: service.litellmPrefix,
            hardcodedModels: service.hardcodedModels,
            defaultModelSource: service.defaultModelSource,
            endpointRules: service.endpointRules
        )
    }
}

File: linux/data/AppService.cpp
#include "AppService.h"

// ---------------------------------------------------------------------------
// ApiFormat conversion
// ---------------------------------------------------------------------------

QString apiFormatToString(ApiFormat f) {
    switch (f) {
    case ApiFormat::OpenAiCompatible: return QStringLiteral("OPENAI_COMPATIBLE");
    case ApiFormat::Anthropic:        return QStringLiteral("ANTHROPIC");
    case ApiFormat::Google:           return QStringLiteral("GOOGLE");
    }
    return QStringLiteral("OPENAI_COMPATIBLE");
}

ApiFormat apiFormatFromString(const QString &s) {
    if (s == QLatin1String("ANTHROPIC")) return ApiFormat::Anthropic;
    if (s == QLatin1String("GOOGLE"))    return ApiFormat::Google;
    return ApiFormat::OpenAiCompatible;
}

// ---------------------------------------------------------------------------
// EndpointRule
// ---------------------------------------------------------------------------

QJsonObject EndpointRule::toJson() const {
    QJsonObject obj;
    obj[QStringLiteral("modelPrefix")]  = modelPrefix;
    obj[QStringLiteral("endpointType")] = endpointType;
    return obj;
}

EndpointRule EndpointRule::fromJson(const QJsonObject &obj) {
    EndpointRule rule;
    rule.modelPrefix  = obj[QStringLiteral("modelPrefix")].toString();
    rule.endpointType = obj[QStringLiteral("endpointType")].toString();
    return rule;
}

// ---------------------------------------------------------------------------
// AppService - static registry
// ---------------------------------------------------------------------------

QList<AppService> AppService::entries;

AppService *AppService::findById(const QString &id) {
    for (auto &s : entries) {
        if (s.id == id) return &s;
    }
    return nullptr;
}

// ---------------------------------------------------------------------------
// AppService - modelFilterRegex (lazy cached)
// ---------------------------------------------------------------------------

const QRegularExpression *AppService::modelFilterRegex() const {
    if (!modelFilterRegex_.has_value()) {
        if (modelFilter.has_value() && !modelFilter->isEmpty()) {
            modelFilterRegex_ = QRegularExpression(
                *modelFilter,
                QRegularExpression::CaseInsensitiveOption
            );
        } else {
            // Store an invalid regex to signal "no filter" — checked via has_value + isValid
            modelFilterRegex_ = QRegularExpression();
        }
    }
    const QRegularExpression &re = *modelFilterRegex_;
    if (re.pattern().isEmpty()) return nullptr;
    return &re;
}

// ---------------------------------------------------------------------------
// ProviderDefinition - JSON deserialization
// ---------------------------------------------------------------------------

static std::optional<QString> optString(const QJsonObject &obj, const QString &key) {
    if (!obj.contains(key) || obj[key].isNull()) return std::nullopt;
    return obj[key].toString();
}

static std::optional<double> optDouble(const QJsonObject &obj, const QString &key) {
    if (!obj.contains(key) || obj[key].isNull()) return std::nullopt;
    return obj[key].toDouble();
}

static std::optional<QStringList> optStringList(const QJsonObject &obj, const QString &key) {
    if (!obj.contains(key) || !obj[key].isArray()) return std::nullopt;
    QStringList list;
    const QJsonArray arr = obj[key].toArray();
    for (const auto &v : arr) {
        list.append(v.toString());
    }
    return list;
}

ProviderDefinition ProviderDefinition::fromJson(const QJsonObject &obj) {
    ProviderDefinition pd;
    pd.id              = obj[QStringLiteral("id")].toString();
    pd.displayName     = obj[QStringLiteral("displayName")].toString();
    pd.baseUrl         = obj[QStringLiteral("baseUrl")].toString();
    pd.adminUrl        = obj[QStringLiteral("adminUrl")].toString();
    pd.defaultModel    = obj[QStringLiteral("defaultModel")].toString();

    pd.openRouterName     = optString(obj, QStringLiteral("openRouterName"));
    pd.chatPath           = optString(obj, QStringLiteral("chatPath"));
    pd.modelsPath         = optString(obj, QStringLiteral("modelsPath"));
    pd.prefsKey           = optString(obj, QStringLiteral("prefsKey"));
    pd.seedFieldName      = optString(obj, QStringLiteral("seedFieldName"));
    pd.modelFilter        = optString(obj, QStringLiteral("modelFilter"));
    pd.litellmPrefix      = optString(obj, QStringLiteral("litellmPrefix"));
    pd.defaultModelSource = optString(obj, QStringLiteral("defaultModelSource"));
    pd.modelListFormat    = optString(obj, QStringLiteral("modelListFormat"));

    pd.apiFormat = obj[QStringLiteral("apiFormat")].toString(QStringLiteral("OPENAI_COMPATIBLE"));

    pd.supportsCitations    = obj[QStringLiteral("supportsCitations")].toBool(false);
    pd.supportsSearchRecency = obj[QStringLiteral("supportsSearchRecency")].toBool(false);
    pd.extractApiCost       = obj[QStringLiteral("extractApiCost")].toBool(false);

    pd.costTicksDivisor = optDouble(obj, QStringLiteral("costTicksDivisor"));
    pd.hardcodedModels  = optStringList(obj, QStringLiteral("hardcodedModels"));

    // Endpoint rules
    if (obj.contains(QStringLiteral("endpointRules")) && obj[QStringLiteral("endpointRules")].isArray()) {
        const QJsonArray arr = obj[QStringLiteral("endpointRules")].toArray();
        for (const auto &v : arr) {
            pd.endpointRules.append(EndpointRule::fromJson(v.toObject()));
        }
    }

    return pd;
}

// ---------------------------------------------------------------------------
// ProviderDefinition - JSON serialization
// ---------------------------------------------------------------------------

QJsonObject ProviderDefinition::toJson() const {
    QJsonObject obj;
    obj[QStringLiteral("id")]          = id;
    obj[QStringLiteral("displayName")] = displayName;
    obj[QStringLiteral("baseUrl")]     = baseUrl;
    if (!adminUrl.isEmpty())
        obj[QStringLiteral("adminUrl")] = adminUrl;
    obj[QStringLiteral("defaultModel")] = defaultModel;

    if (openRouterName.has_value())
        obj[QStringLiteral("openRouterName")] = *openRouterName;
    if (apiFormat != QLatin1String("OPENAI_COMPATIBLE"))
        obj[QStringLiteral("apiFormat")] = apiFormat;
    if (chatPath.has_value())
        obj[QStringLiteral("chatPath")] = *chatPath;
    if (modelsPath.has_value())
        obj[QStringLiteral("modelsPath")] = *modelsPath;
    if (prefsKey.has_value())
        obj[QStringLiteral("prefsKey")] = *prefsKey;
    if (seedFieldName.has_value() && *seedFieldName != QLatin1String("seed"))
        obj[QStringLiteral("seedFieldName")] = *seedFieldName;
    if (supportsCitations)
        obj[QStringLiteral("supportsCitations")] = true;
    if (supportsSearchRecency)
        obj[QStringLiteral("supportsSearchRecency")] = true;
    if (extractApiCost)
        obj[QStringLiteral("extractApiCost")] = true;
    if (costTicksDivisor.has_value())
        obj[QStringLiteral("costTicksDivisor")] = *costTicksDivisor;
    if (modelListFormat.has_value() && *modelListFormat != QLatin1String("object"))
        obj[QStringLiteral("modelListFormat")] = *modelListFormat;
    if (modelFilter.has_value())
        obj[QStringLiteral("modelFilter")] = *modelFilter;
    if (litellmPrefix.has_value())
        obj[QStringLiteral("litellmPrefix")] = *litellmPrefix;
    if (defaultModelSource.has_value())
        obj[QStringLiteral("defaultModelSource")] = *defaultModelSource;

    if (hardcodedModels.has_value()) {
        QJsonArray arr;
        for (const auto &m : *hardcodedModels) {
            arr.append(m);
        }
        obj[QStringLiteral("hardcodedModels")] = arr;
    }

    if (!endpointRules.isEmpty()) {
        QJsonArray arr;
        for (const auto &rule : endpointRules) {
            arr.append(rule.toJson());
        }
        obj[QStringLiteral("endpointRules")] = arr;
    }

    return obj;
}

// ---------------------------------------------------------------------------
// ProviderDefinition <-> AppService conversion
// ---------------------------------------------------------------------------

AppService ProviderDefinition::toAppService() const {
    AppService s;
    s.id          = id;
    s.displayName = displayName;
    s.baseUrl     = baseUrl;
    s.adminUrl    = adminUrl;
    s.defaultModel = defaultModel;
    s.openRouterName = openRouterName;
    s.apiFormat   = apiFormatFromString(apiFormat);
    s.chatPath    = chatPath.value_or(QStringLiteral("v1/chat/completions"));
    s.modelsPath  = modelsPath;
    s.prefsKey    = prefsKey.value_or(id.toLower());
    s.seedFieldName = seedFieldName.value_or(QStringLiteral("seed"));
    s.supportsCitations    = supportsCitations;
    s.supportsSearchRecency = supportsSearchRecency;
    s.extractApiCost       = extractApiCost;
    s.costTicksDivisor     = costTicksDivisor;
    s.modelListFormat = modelListFormat.value_or(QStringLiteral("object"));
    s.modelFilter        = modelFilter;
    s.litellmPrefix      = litellmPrefix;
    s.hardcodedModels    = hardcodedModels;
    s.defaultModelSource = defaultModelSource;
    s.endpointRules      = endpointRules;
    return s;
}

ProviderDefinition ProviderDefinition::fromAppService(const AppService &s) {
    ProviderDefinition pd;
    pd.id              = s.id;
    pd.displayName     = s.displayName;
    pd.baseUrl         = s.baseUrl;
    pd.adminUrl        = s.adminUrl;
    pd.defaultModel    = s.defaultModel;
    pd.openRouterName  = s.openRouterName;
    pd.apiFormat       = apiFormatToString(s.apiFormat);
    pd.chatPath        = s.chatPath;
    pd.modelsPath      = s.modelsPath;
    pd.prefsKey        = s.prefsKey;
    pd.seedFieldName   = s.seedFieldName;
    pd.supportsCitations    = s.supportsCitations;
    pd.supportsSearchRecency = s.supportsSearchRecency;
    pd.extractApiCost       = s.extractApiCost;
    pd.costTicksDivisor     = s.costTicksDivisor;
    pd.modelListFormat = s.modelListFormat;
    pd.modelFilter        = s.modelFilter;
    pd.litellmPrefix      = s.litellmPrefix;
    pd.hardcodedModels    = s.hardcodedModels;
    pd.defaultModelSource = s.defaultModelSource;
    pd.endpointRules      = s.endpointRules;
    return pd;
}
